### ISSUE TITLE
Cleanup specs and structure

### ISSUE THREAD

**Issue by DanielOaks** [ircv3/ircv3-specifications#329]
This is something I've been meaning to do for a while, just making an issue so I remember to submit the changes. Once #327 is done and SASL is merged I'll do this.

1. Remove version numbers from spec filenames and YAML title names.
2. Where they exist, remove markdown headers that duplicate the YAML title headers on specs (some specs for instance have an `## IRCv3 <blah> Extension` up the top in addition to the YAML header – rip that out.
3. Reorganise the specs under a few separate folders.

Since this repo's a submodule named `/specs` on the website, makes sense to have a few subfolders:

- `specs/client-tags`: Client tag specifications.
- `specs/batches`: Batch type specifications.
- `specs/sasl-mechs`: SASL-mechanism-specific specs.
- `specs/metadata-keys`: Metadata key specs.
- `specs/deprecated`: Deprecated specs we want to keep around (such as the current SASL mech specs under `specs/documentation`).
- `specs/extensions`: Everything not covered by the above.

Yep, under this arrangement everything currently in `specs/core` gets put into `specs/extensions` instead. It makes sure people know that things are optional (things like metadata aren't mandatory to implement to be IRCv3-compliant), and the main [Specs page](http://ircv3.net/irc/) on the website is better positioned to make the necessary separation between, for example, cap, message-tags, and specs like extended-join anyway.

**Comment by DanielOaks** [comment]
Note to self when SASL gets merged, fit this example into the merged spec:

-----

Example showing new SASL mechanisms becoming available. This example requires the client to support `CAP LS 302`:		

        Client: CAP LS 302
        Server: :irc.example.com CAP * LS :sasl=PLAIN
        ...
        Server: :irc.example.com CAP client DEL :sasl
        Server: :irc.example.com CAP client NEW :sasl=PLAIN,EXTERNAL
		
The given `NEW`/`DEL` messages may be displayed together for simplicity's sake.

**Comment by jwheare** [comment]
We could also note that 908 can be used to advertise a change of mechlist

**Comment by DanielOaks** [comment]
For changing mechanisms, should we push CAP DEL+NEW or 908? Maybe if they've already negotiated the `sasl` cap and the mechanism list changes, just send them 908 yeah?

**Comment by jwheare** [comment]
We could push 908 for changed mechlists but still have an example of sasl going away and then coming back on e.g. a services outage. Clients should handle both.

**Comment by DanielOaks** [comment]
Yep, that sounds like a good way to split them up. I'll do it that way in the SASL merge then \o/

**Comment by jwheare** [comment]
As mentioned on IRC. This should probably be a MUST:
> If the client completes registration (with CAP END, NICK, USER and any other necessary messages) while the SASL authentication is still in progress, the server SHOULD abort it and send a 906 numeric, then register the client without authentication.

**Comment by grawity** [comment]
As long as the MUST phrasing doesn't accidentally forbid the server from rejecting unauthenticated clients

**Comment by DanielOaks** [comment]
After the SASL specs get merged, we can throw another PR in that actually makes changes. Putting material changes in with the merge slows stuff down.

**Comment by jwheare** [comment]
Fine by me!

**Comment by DanielOaks** [comment]
Message-tags has also now been merged into a single spec. Whoo!

**Comment by DanielOaks** [comment]
Reorganised the list of things-to-be-done. Much more action-oriented now.

- 2/3/4) Pretty simple, just going through and changing and unifying file formats, so no biggie I'll get to it pretty soon.
- 5) Merging SASL is a pretty big job, my time merging CAP taught me that there's a lot to keep in mind. It's a much shorter spec though so hopefully nowhere near as complicated. Don't touch SASL specs in 2+3+4 changes, since it's got an open nontrivial PR into it and all.
- 6) Examples are hard. This'll be a spec-by-spec process. May involve some thoughts around how we display examples generally on the site / in our specs.
- 7) Future me wants to do this. Current me does not and only sees pain down this path. Good luck to the future IRCv3 team of 2070 who manages to work on this.
